arbor/simulation.cpp

```diff
@@ -558,10 +565,41 @@ void simulation::set_local_spike_callback(spike_export_function export_callback)
     impl_->local_export_callback_ = std::move(export_callback);
 }
 
+void simulation::set_epoch_callback(epoch_function epoch_callback, bool global) {
+    impl_->epoch_callback_ = std::move(epoch_callback);
+}
+
 void simulation::inject_events(const cse_vector& events) {
     impl_->inject_events(events);
 }
 
 simulation::~simulation() = default;
 
+epoch_function epoch_progress_bar() {
+    struct impl {
+        double t0 = 0;
+        unsigned calls = 0;
+
+        const char* PBSTR = "||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||";
+        unsigned PBWIDTH = 50;
+
+        void operator() (double t, double tfinal) {
+            if (calls==0) {
+                t0 = t;
+            }
+
+            double percentage = (tfinal==t0)? 1: (t-t0)/(tfinal-t0);
+            int val = percentage * 100;
+            int lpad = percentage * PBWIDTH;
+            int rpad = PBWIDTH - lpad;
+            printf("\r%3d%% [%.*s%*s]  %12ums", val, lpad, PBSTR, rpad, "", (unsigned)t);
+            fflush(stdout);
```

Comment:
:+1: 

I tweaked it to print a new line and reset the counter so that code like this:
```python
sim.run(1000, 0.1)
sim.run(2000, 0.1)
sim.run(3000, 0.1)
```
will generate the following output:
```
100% |--------------------------------------------------|          1000ms
100% |--------------------------------------------------|          2000ms
100% |--------------------------------------------------|          3000ms
```